web/modules/custom/digital_gov_migration/digital_gov_migration.module

```diff
@@ -0,0 +1,112 @@
+<?php
+
+/**
+ * @file
+ * Primary module hooks for Digital.gov Migration module.
+ */
+
+use Drupal\Core\Entity\EntityInterface;
+use Drupal\redirect\Entity\Redirect;
+
+/**
+ * Implements hook_entity_insert().
+ */
+function digital_gov_migration_entity_insert(EntityInterface $entity) {
+  // Check if the entity is a node.
+  if ($entity->getEntityTypeId() !== 'node') {
+    return;
+  }
+  if (!empty($entity->redirects)) {
+    // Create redirects on nodes as they are migrated. If the migration is
+    // rolled back, the redirects are removed automatically by Drupal.
+    // This is not needed once the migration is over.
+    $nid = $entity->id();
+    $node_url = '/node/' . $nid;
+
+    foreach ($entity->redirects as $redirect) {
+      if (empty($redirect['redirect_source']['path'])) {
+        throw new \Exception('Cannot save redirect without a source/path ' . $node_url);
+      }
+      $redirect['redirect_source']['path'] = ltrim($redirect['redirect_source']['path'], '/');
+      // Create a redirect entity.
+      $redirect = Redirect::create([
+        'redirect_source' => $redirect['redirect_source'],
+        'redirect_redirect' => ['uri' => 'internal:' . $node_url],
+        'status_code' => $redirect['status_code'] ?? 301,
+      ]);
+
+      // Save the redirect.
+      $redirect->save();
+    }
+  }
+
+}
+
+/**
+ * Implements hook_entity_presave().
+ */
+function digital_gov_migration_entity_presave(EntityInterface $entity) {
+  // Check if the entity is a node.
+  if ($entity->getEntityTypeId() !== 'node') {
+    return;
+  }
+  // Venue only exists during the migration, this wouldn't effect updating
+  // existing items.
+  if (!empty($entity->venue)) {
+    $values = [
+      'type' => 'venue',
+      'field_venue_address' => $entity->venue['address'] ?? '',
+      'field_venue_city' => $entity->venue['city'] ?? '',
+      'field_venue_country' => $entity->venue['country'] ?? '',
+      'field_venue_map' => $entity->venue['map'] ?? '',
+      'field_venue_name' => $entity->venue['venue_name'] ?? '',
+      'field_venue_room' => $entity->venue['room'] ?? '',
+      'field_venue_state' => $entity->venue['state'] ?? '',
+      'field_venue_zip' => $entity->venue['zip'] ?? '',
+    ];
+    /** @var \Drupal\paragraphs\Entity\Paragraph $venue */
+    $venue = \Drupal::entityTypeManager()->getStorage('paragraph')->create($values);
+    $venue->save();
+    $entity->get('field_venue')->appendItem($venue);
+  }
+}
+
+/**
+ * Implements hook_entity_delete().
+ */
+function digital_gov_migration_entity_delete(EntityInterface $entity) {
+  // Clean up paragraphs that cannot be rolled back because they were not part
+  // of a migration.
+  $is_a_rollback = FALSE;
+  foreach (debug_backtrace() as $item) {
+    if ($item['function'] ?? '' === 'rollback') {
+      $is_a_rollback = TRUE;
+    }
+  }
+  if (!$is_a_rollback || !$entity->hasField('field_venue') || $entity->get('field_venue')->isEmpty()) {
```

Comment:
`lando si` keeps running into a fatal error for me here. It's getting an `Entity` instance that doesn't have a `hasField()` method. Can you add this check to the chain of ORs here?

```php
!($entity instanceof \Drupal\paragraphs\Entity\Paragraph)
```

Reply:
I've updated the logic here so that it does 'easier' tests first and fixed a logical error with my equality. Thanks @omerida !